### PR TITLE
#161254661 Implement social login

### PR DESCRIPTION
**Pull request by evamaina**
#### What does this PR do?
```This PR enables a user to login to Authors Haven via social media accounts i.e. Twitter, Facebook, Google ```
#### Description of Task to be completed?
```Have social login  endpoint working```

#### How should this be manually tested?

 - For Google, get the access token for oauth2 from the [Google Developer Playground](https://developers.google.com/oauthplayground/) and use it as follows:
```
 {
  "access_token": "some-access-token",
  "provider":"google-oauth2"
} 
```
- For Twitter, create an app on the [Twitter Developer Platform](https://developer.twitter.com/) to get the consumer key and secret and use it as follows:
```
{
  "access_token": "your-consumer-key",
   "access_token_secret": "your-consumer-secret",
  "provider":"twitter"
} 
```

*  For Facebook, create an app on the [Facebook Developer Platform](https://developers.facebook.com/apps/) to get the facebook key and secret and use it as follows:
```
{
  "access_token": "your-consumer-key",
  "provider":"facebook"
} 
```
#### Any background context you want to provide?
```Note that when using twitter authentication,apart from "access_token" you also need to provide "access_token_secret" because it uses oauth1```

#### What are the relevant pivotal tracker stories?
[#161254661](https://www.pivotaltracker.com/n/projects/2205267/stories/161254661) 

#### Screenshots 
image

image

image

### CODE REVIEW

**Comment by abadojack** [comment]
@code-sleuth @kipropbrian Testing google, facebook and twitter auth with valid tokens is really stressing since we have to generate new tokens almost every time we run the tests. As a result, we decided to comment out the tests. However, this results to a drop in test coverage.

File: .travis.yml [hunk 1]
```diff
@@ -1,16 +1,21 @@
 language: python
 python:
   - "3.6"
-services:
-  - postgresql
+
+addons:
+  postgresql: "9.4"
```
Comment (code-sleuth) [inline .travis.yml]: Why are we using postgres `9.4`? I suggest we use `9.6`

File: authors/apps/authentication/serializers.py [hunk 2]
```diff
@@ -108,10 +108,15 @@ class UserSerializer(serializers.ModelSerializer):
         min_length=8,
         write_only=True
     )
+    token = serializers.CharField(
+        max_length=128,
+        read_only=True
+
```
Comment (code-sleuth) [inline authors/apps/authentication/serializers.py]: Remove this space.

File: authors/apps/authentication/views.py [hunk 3]
```diff
@@ -223,3 +229,99 @@ def put(self, request, **kwargs):
...
+    def get_auth_user(self, request, backend, token, provider):
+        # If this request is made with an authenticated user,
+        #  try to associate a social account with it
+        authed_user = request.user if not request.user.is_anonymous else None
+
+        try:
+            # if `authed_user` is None,
+            #   python-social-auth will make a new user,
+            # else
+            #   this social account will be associated with the user passed
```
Comment (code-sleuth) [inline authors/apps/authentication/views.py]: It should be `...user parsed.`
Context:
`Parsing` means trying to interpret an object and break it down into data elements of some kind while `Passing` is when you give parameters to a method.

In this scenario you are interpreting the `user` from the `token` and `authed_user` variables so what you are doing is `parsing`.

File: authors/apps/authentication/views.py [hunk 4]
```diff
@@ -223,3 +229,99 @@ def put(self, request, **kwargs):
...
+            user = backend.do_auth(token, user=authed_user)
+        except AuthAlreadyAssociated:
+            return Response({
+                "errors": "That social media account is already in use"
+            }, status=status.HTTP_400_BAD_REQUEST)
```
Comment (code-sleuth) [inline authors/apps/authentication/views.py]: This should be a conflict status code returned. `409`

**Comment by abadojack** [comment]
@code-sleuth https://groups.google.com/forum/#!topic/api-craft/ULgaVic7OCU according to this post, after sign in, a user is required to authorise access to their data before one can get the token. We cannot do that in code. The same applies for facebook

File: .travis.yml (hunk 1 above)
Comment (abadojack) [reply .travis.yml]: Ok. I have used 9.6

File: authors/apps/authentication/serializers.py (hunk 2 above)
Comment (abadojack) [reply authors/apps/authentication/serializers.py]: Done

File: authors/apps/authentication/views.py (hunk 3 above)
Comment (abadojack) [reply authors/apps/authentication/views.py]: OK

File: authors/apps/authentication/views.py (hunk 4 above)
Comment (abadojack) [reply authors/apps/authentication/views.py]: Done. Changed to 409

**Comment by abadojack** [comment]
@code-sleuth I've made the changes you requested

**Comment by Eyansky** [comment]
LGTM ... but coverage issue😥

**Pull request merged**